### PR TITLE
Add `http_parser` library dependency

### PR DESCRIPTION
**Pull request by staticfloat**
Tested to be building `libhttp_parser` and using it with `libgit2` on OSX and Linux.

From what I can tell of the `libgit2` build system, it will not use a system-provided `http_parser` on Windows, but this PR will build it anyway.

### CODE REVIEW

File: deps/dsfmt.mk [hunk 1]
```diff
@@ -43,8 +43,7 @@ define DSFMT_INSTALL
 endef
 $(eval $(call staged-install, \
 	dsfmt,dsfmt-$(DSFMT_VER), \
-	DSFMT_INSTALL,, \
-	$$(DSFMT_OBJ_TARGET), \
+	DSFMT_INSTALL,,, \
```
Comment (tkelman) [inline deps/dsfmt.mk]: looks unrelated?

File: deps/http_parser.mk [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	-rm -f $(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/build-compiled $(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/build-checked $(build_shlibdir)/libhttp-parser.$(SHLIB_EXT)
+	-$(MAKE) -C $(BUILDDIR)/$(HTTPPARSER_SRC_DIR) $(HTTPPARSER_FLAGS) clean
+
+
+get-http_parser: $(HTTPPARSER_SRC_FILE)
+extract-http_parser: $(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/source-extracted
+configure-http_parser: extract-http_parser
+compile-http_parser: $(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/build-compiled
+fastcheck-http_parser: check-http_parser
+check-http_parser: compile-http_parser
```
Comment (tkelman) [inline deps/http_parser.mk]: does it have unit tests?

File: deps/http_parser.version [hunk 3]
```diff
@@ -0,0 +1,2 @@
+HTTPPARSER_BRANCH=master
```
Comment (tkelman) [inline deps/http_parser.version]: use a release tag

File: deps/http_parser.mk [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+# Also manually set PREFIX to DESTDIR
+HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_prefix)
+
+$(eval $(call staged-install, \
+	http_parser,$$(HTTPPARSER_SRC_DIR), \
+	MAKE_INSTALL,$(HTTPPARSER_FLAGS),, \
+	$(INSTALL_NAME_CMD)libhttp_parser.$(SHLIB_EXT) $(build_shlibdir)/libhttp_parser.$(SHLIB_EXT)))
+
+$(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/http_parser-makefile.patch-applied: $(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/source-extracted
+	cd $(BUILDDIR)/$(HTTPPARSER_SRC_DIR) && patch -p1 -f < $(SRCDIR)/patches/http_parser-makefile.patch
```
Comment (tkelman) [inline deps/http_parser.mk]: patching git-external deps ends up causing issues. we should upstream any relevant parts of this, or set make flags for others

File: deps/patches/http_parser-makefile.patch [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
++	$(STRIP) $(DESTDIR)$(LIBDIR)/$(SONAME)
+ 
+ uninstall:
+-	rm $(INCLUDEDIR)/http_parser.h
+-	rm $(LIBDIR)/$(SONAME)
+-	rm $(LIBDIR)/libhttp_parser.so
++	rm $(DESTDIR)$(INCLUDEDIR)/http_parser.h
++	rm $(DESTDIR)$(LIBDIR)/$(SONAME)
++	rm $(DEstdiR)$(LIBDIR)/libhttp_parser.$(SOEXT)
++	rm $(DEstdiR)$(LIBDIR)/libhttp_parser.a
```
Comment (tkelman) [inline deps/patches/http_parser-makefile.patch]: what's with the capitalization here?

File: deps/patches/http_parser-makefile.patch [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ 
+ clean:
+ 	rm -f *.o *.a tags test test_fast test_g \
+ 		http_parser.tar libhttp_parser.so.* \
+ 		url_parser url_parser_g parsertrace parsertrace_g \
+-		*.exe *.exe.so
++		*.exe.so *.dll
```
Comment (tkelman) [inline deps/patches/http_parser-makefile.patch]: exe should probably still be cleaned

**Review by tkelman — CHANGES_REQUESTED** [review]
should also go in license and readme

how hard is it to make libgit2 use an external http-parser library on windows?

File: deps/http_parser.mk (hunk 2 above)
Comment (staticfloat) [inline deps/http_parser.mk]: Heh, I went through the trouble of adding the `$(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/build-checked` recipe, but then didn't make it a target of anything.  :P

File: deps/dsfmt.mk (hunk 1 above)
Comment (staticfloat) [inline deps/dsfmt.mk]: Yeah, this is a separate commit.  I'll move it out for now.

File: deps/http_parser.mk (hunk 4 above)
Comment (staticfloat) [inline deps/http_parser.mk]: Upstream has had something similar to this patch open as a [PR since June](https://github.com/nodejs/http-parser/pull/321), it looks like upstream is a little slow to respond to PRs.  The patch carried here is based off of the patch in that PR, with some extra stuff for Julia.  I'll see if I can pare the patch down, I was just copying it wholesale from wildart's original work.

File: deps/patches/http_parser-makefile.patch (hunk 5 above)
Comment (staticfloat) [inline deps/patches/http_parser-makefile.patch]: This was part of the original patch in the main http-parser PR, I'll clean this patch up.

File: deps/patches/http_parser-makefile.patch (hunk 6 above)
Comment (staticfloat) [inline deps/patches/http_parser-makefile.patch]: Agreed.

**Comment by staticfloat** [comment]
Alright, I think I've addressed all comments so far.  I have no idea what it takes to make libgit2 use a system-provided `http_parser`, looking at the [cmake file](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L238-L241), it looks like if we use `WinHTTP` as the transport layer, we have no choice but to build the libgit2-provided `http_parser`.  As [this is the default](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L83) and we don't seem to override that anywhere in our building of `libgit2`, I'd say it's currently impossible to do so.

I don't know what the advantages/disadvantages of using `WinHTTP` are over "the built-in HTTP transport", so I don't know if changing that libgit2 option is a good idea or not.

File: deps/Makefile [hunk 7]
```diff
@@ -22,7 +22,7 @@ include $(SRCDIR)/tools/git-external.mk
 # autoconf configure-driven scripts: llvm pcre arpack fftw unwind gmp mpfr patchelf libuv curl
 # custom Makefile rules: openlibm dsfmt suitesparse-wrapper suitesparse lapack openblas utf8proc objconv osxunwind
 # entirely custom: virtualenv
-# CMake libs: libgit2 libssh2 mbedtls
+# CMake libs: http_parser libgit2 libssh2 mbedtls
```
Comment (tkelman) [inline deps/Makefile]: it doesnt use cmake...

File: deps/http_parser.mk [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+HTTPPARSER_GIT_URL := git://github.com/nodejs/http-parser.git
+HTTPPARSER_TAR_URL = https://api.github.com/repos/nodejs/http-parser/tarball/$1
+$(eval $(call git-external,http_parser,HTTPPARSER,,,$(BUILDDIR)))
+
+# Manually set the PLATFORM flag to whatever $(OS) is, because at least one
+# person (@staticfloat) has the PLATFORM envvar set to something else.
+# Also manually set PREFIX to DESTDIR
+HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_prefix)
+
+$(eval $(call staged-install, \
```
Comment (tkelman) [inline deps/http_parser.mk]: does this create the manifest and uninstall targets?

Comment (staticfloat) [inline deps/http_parser.mk]: [Yes](https://github.com/JuliaLang/julia/blob/548465b6ded7d2d4f908760e66329d7fcdfd356d/deps/tools/common.mk#L159-L163) and [Yes](https://github.com/JuliaLang/julia/blob/548465b6ded7d2d4f908760e66329d7fcdfd356d/deps/tools/common.mk#L132-L134)

File: deps/Makefile (hunk 7 above)
Comment (staticfloat) [inline deps/Makefile]: doh

**Comment by staticfloat** [comment]
@tkelman is this all good to go from your viewpoint?

**Comment by tkelman** [comment]
Has libgit2 modified their copy of http-parser to work with winhttp?

Should also trial run this through the buildbots first.

**Comment by staticfloat** [comment]
> Has libgit2 modified their copy of http-parser to work with winhttp?

It looks to me like [winhttp is on by default](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L80-L84), so I'd say yes.

> Should also trial run this through the buildbots first.

[macOS build](https://build.julialang.org/builders/package_osx10.9-x64/builds/1015): Build finished, `libhttp_parser.2.7.1.dylib` confirmed to be in the archive.
[linux32 build](https://build.julialang.org/builders/package_tarball32/builds/998): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[linux64 build](https://build.julialang.org/builders/package_tarball64/builds/1032): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[arm build](https://build.julialang.org/builders/package_tarballarm/builds/974): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[ppc64le build](https://build.julialang.org/builders/package_tarballppc64le/builds/234): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[win32 build](https://build.julialang.org/builders/package_win6.2-x86/builds/887): Build finished, I can see that `libhttp_parser.2.7.1.dll` has been built, and I assume it is also in the archive, but I can't see that just from the logs.
[win64 build](https://build.julialang.org/builders/package_win6.2-x64/builds/958): Build finished, I can see that `libhttp_parser.2.7.1.dll` has been built, and I assume it is also in the archive, but I can't see that just from the logs.

**Comment by tkelman** [comment]
I meant the source code. If it's a matter of adding a define when building http-parser, we can easily do that ourselves.

File: deps/patches/http_parser-makefile.patch [hunk 9]
```diff
@@ -0,0 +1,98 @@
+diff --git a/Makefile b/Makefile
+index b2476db..92a84c9 100644
+--- a/Makefile
++++ b/Makefile
+@@ -24,6 +24,9 @@ BINEXT ?=
+ ifeq (darwin,$(PLATFORM))
+ SONAME ?= libhttp_parser.2.7.1.dylib
+ SOEXT ?= dylib
++else ifeq (winnt,$(PLATFORM))
++SONAME ?= libhttp_parser-2.7.dll
```
Comment (tkelman) [inline deps/patches/http_parser-makefile.patch]: why was this not the soname on the windows buildbot?

Comment (staticfloat) [inline deps/patches/http_parser-makefile.patch]: Errant brain interpolation

Comment (staticfloat) [inline deps/patches/http_parser-makefile.patch]: Edited and fixed above

**Comment by staticfloat** [comment]
> I meant the source code. If it's a matter of adding a define when building http-parser, we can easily do that ourselves.

If you mean you want to normalize across platforms and _not_ use WinHTTP, then I think all we have to do is to add `-DWINHTTP=OFF` to [this list of libgit2 CMake options](https://github.com/JuliaLang/julia/blob/f7a611eb1ab00d4eef9a50a3a1f09fc1a8ded6e3/deps/libgit2.mk#L27).  I just don't know what the benefit of doing/not doing that is, so I stayed away from it.

**Comment by tkelman** [comment]
winhttp means the windows native crypto backend. We want to use that for https support (we could try using mbedtls instead but that's still fairly experimental, best to keep it linux only). That shouldn't make a huge difference to whether libgit2 builds its vendored copy of http-parser or links to an external copy, unless they've made source changes that are not upstreamed.

**Comment by staticfloat** [comment]
Great, then I think we're ready to go.  As it stands right now, `http_parser` is always built separately (so it's still available for use on Windows), but `libgit2` does not use it on Windows, since WinHTTP is enabled by default.

**Comment by tkelman** [comment]
>  if we use WinHTTP as the transport layer, we have no choice but to build the libgit2-provided http_parser

does libgit2 use http-parser on windows or not? if yes, then we should link against the standalone copy

**Comment by staticfloat** [comment]
Ah, I finally understand what you're asking.  It doesn't look like they have support for using _external_ http-parser on windows.  They [force the usage](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L239-L241) of their internal copy, but [their internal copy](https://github.com/libgit2/libgit2/blob/master/deps/http-parser/http_parser.c) has no mention of windows `#ifdef`'s or whatnot, and doesn't use the `GIT_WINHTTP` preprocessor macro that is defined when WinHTTP is used, so I theorize we should be able to link in our own http-parser if we wish.  We just need to mess around with the CMakeLists file a bit.

**Comment by staticfloat** [comment]
I've made a PR to upstream [here](https://github.com/libgit2/libgit2/pull/4015), we'll see what they think of it.

**Comment by quinnj** [comment]
Bump, are we waiting on something upstream?

**Comment by tkelman** [comment]
Cleaning the PR up and/or incorporating the patch here

**Comment by staticfloat** [comment]
I've added the patch here so we can at least ensure that it works nicely for us.  I'm running it through the buildbots, here are the builds, I'll post links to the binaries so that users such as @quinnj can ensure that they work:

[OSX build](https://build.julialang.org/builders/package_osx10.9-x64/builds/1059): Pending
[Linux x86 build](https://build.julialang.org/builders/package_tarball32/builds/1043): Pending
[Linux x64 build](https://build.julialang.org/builders/package_tarball64/builds/1074): Pending
[Linux ARM build](https://build.julialang.org/builders/package_tarballarm/builds/1016): Pending
[Linux PPC64le build](https://build.julialang.org/builders/package_tarballppc64le/builds/276): Pending
[Windows x64 build](https://build.julialang.org/builders/package_win6.2-x64/builds/1000): Pending
[Windows x86 build](https://build.julialang.org/builders/package_win6.2-x86/builds/933): Pending

**Comment by staticfloat** [comment]
The `.dll` file isn't getting put in the right place on Windows; it's being put into `/lib` instead of `/bin`.  I've pushing the latest commit (d15edf2) through the windows builders, we'll see if this fixes things up.  In the meantime, the earlier commit worked on OSX and Linux 64:

```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser"))
"/tmp/julia-121f07d151/bin/../lib/julia/libhttp_parser.so.2.7.1"
```

```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser"))
"/Users/sabae/Desktop/Julia-0.6.app/Contents/Resources/julia/lib/julia//libhttp_parser.dylib"
```

**Comment by staticfloat** [comment]
New binaries available:

Win32 [download link](https://s3.amazonaws.com/julianightlies/bin/winnt/x86/0.6/julia-0.6.0-d15edf2f3b-win32.exe).
Win64 [download link](https://s3.amazonaws.com/julianightlies/bin/winnt/x64/0.6/julia-0.6.0-d15edf2f3b-win64.exe).

Things are not working.  First off, the non-versioned filename appears to be some mishmash of shortcut and dll.  I'm not sure what's going on there, but trying to `dlopen` it doesn't work.

Secondly, the versioned filename is not loading properly.  Looking at it with DepWalker reveals a dependency on `cygwin1.dll`.  Is that......normal?

**Comment by tkelman** [comment]
That means it used the wrong toolchain, it wasn't correctly cross compiled.

**Comment by staticfloat** [comment]
Alright, serious progress.  With [this binary](https://s3.amazonaws.com/julianightlies/bin/winnt/x64/0.6/julia-0.6.0-6e1377f2a4-win64.exe), I now get:
```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser-2.7.dll"))
"C:\\Users\\Elliot Saba\\AppData\\Local\\Julia-0.6.0-dev\\bin\\libhttp_parser-2.7.dll"
```

The only remaining problem is the weird `libhttp_parser.dll` file.  @tkelman what's SOP on windows; install only the versioned filename and use that explicitly within `dlopen()` calls?  Install both?  Install only unversioned filename?

**Comment by tkelman** [comment]
If you put a version name in the dll, then should not use the unversioned name at all. Unless you're building for cygwin, which we aren't.

File: deps/http_parser.mk [hunk 10]
```diff
@@ -7,7 +7,7 @@ $(eval $(call git-external,http_parser,HTTPPARSER,,,$(BUILDDIR)))
 # Manually set the PLATFORM flag to whatever $(OS) is, because at least one
 # person (@staticfloat) has the PLATFORM envvar set to something else.
 # Also manually set PREFIX to DESTDIR
-HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_prefix)
+HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_prefix) LIBDIR=$(build_shlibdir)
```
Comment (staticfloat) [inline deps/http_parser.mk]: My feeling is that they didn't really have windows in mind when they designed this.

Comment (tkelman) [inline deps/http_parser.mk]: if they're putting anything other than the dll there, thrn it's not right to put it in bin

Comment (staticfloat) [inline deps/http_parser.mk]: [They're not](https://github.com/nodejs/http-parser/blob/master/Makefile#L125-L128). 

Could `ln -s` be creating a nasty file on Windows?  Are we not allowed to have symlinks on Windows?  If not, my proposed solution is to just add an `rm libhttp_parser.dll` to the windows install steps and tell everyone to add the fully versioned name to any bindeps scripts that need it.

Comment (tkelman) [inline deps/http_parser.mk]: yes, if run in cygwin then ln -s creates a file that only cygwin can read. if run in msys, ln -s just copies i believe, so it's never really the right thing to do for mingw.

File: deps/http_parser.mk [hunk 11]
```diff
@@ -12,6 +12,7 @@ HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_pre
 $(eval $(call staged-install, \
 	http_parser,$$(HTTPPARSER_SRC_DIR), \
 	MAKE_INSTALL,$(HTTPPARSER_FLAGS),, \
+	[[ "$(OS)" == "WINNT" ]] && mv $(build_shlibdir)/libhttp_parser-2.7.dll $(build_shlibdir)/libhttp_parser.dll; \
```
Comment (staticfloat) [inline deps/http_parser.mk]: Ah, good to know.  Reverted.

**Comment by staticfloat** [comment]
Looks like this last commit gets everything on the up-and-up.  Final windows binaries:

[Win x86](https://s3.amazonaws.com/julianightlies/bin/winnt/x86/0.6/julia-0.6.0-8fdea47fd9-win32.exe)
[Win x64](https://s3.amazonaws.com/julianightlies/bin/winnt/x64/0.6/julia-0.6.0-8fdea47fd9-win64.exe)

You have to access the library by its versioned filename:
```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser-2.7.dll"))
"C:\\Users\\Elliot Saba\\AppData\\Local\\Julia-0.6.0-dev\\bin\\libhttp_parser-2.7.dll"
```

**Comment by tkelman** [comment]
Or update the patch to be consistent with what got merged. Either way I realized this is of limited usefulness since not everyone will be building libgit2 this way, so like with mbedtls the package will need to have a fallback to deal with cases where the libgit2 that Julia uses is built differently.

**Comment by staticfloat** [comment]
With the advent of things like HTTP.jl, I think this is less useful than it would have been a year ago, and I'm just going to close this for now.
